### ISSUE TITLE
Better way to customize equality in unit tests and better failure message

### ISSUE THREAD

**Issue by nevillelyh** [spotify/scio#2368]
Right now most test assertions delegate to Beam's `PAssert` and hamcrest. The most common feedback about writing tests are:

- Hard to apply custom comparison logic, e.g. comparing `List[T]` fields as sets
- Hard to parse failure message, e.g. `containsInAnyOrder` should contain _very long list of records_ but does not ...

For the 1st, cats `Eq[T]` seems like an idiomatic choice but we can't inject it into Beam/hamcrest easily. We either have to hack the internals of these java libs or add our own matchers?

2 is only doable if we add our own matchers instead of using Beam/hamcrest?

**Comment by anish749** [comment]
For number 1:

Scala test supports `NormalizingEquality` using scalactic, however we don't support it.

Our current example is:
```
x:SCollection[T] should containInAnyOrder(expected: Iterable[T])
```
here `containInAnyOrder` returns a `org.scalatest.matchers.Matcher` which gets a bit hacky with Coders as our current signature is `def containInAnyOrder[T: Coder](value: Iterable[T]): IterableMatcher[SCollection[T], T]`
Scalatest supports type classes using MatcherFactory, however we are not using this.

Ideally we want to use `MatcherFactory2[T, Coder, Equality]` instead of `Matcher`, this way we can inject an equality (we need to move away from hamcrest in favor of scalatest to support this)

In that case we can support the syntax:
```
x should containInAnyOrder (Seq("hi", "hello")) (after being convertedToSet)
```

where `(after being convertedToSet)` returns a `NormalizingEquality[T]` for comparing Sequences.

This follows the scalatest and scalatic syntax like:
```
"Hi" should equal ("hi") (after being lowerCased)
```

**Comment by jto** [comment]
Re-implementing our own custom matchers sounds like a lot of works and it might be tricky to handle windowing correctly.

It might be possible to simply wrap each element of the `SCollection` into a wrapper class that overrides `toString` and `equals` ? 

```scala
case class TestWrapper[T: Eq](t: T) {
   override def toString = prettyPrint(this)
   override def equals(o: Any) = Eq[T].equals(t, o.t)
   override def hashCode = ???
}
```

Then matchers like `containInAnyOrder` would start by internally wrapping the elements in `value` and the elements in the tested `SCollection` in `TestWrapper`.

wdyt @nevillelyh @anish749 ?

**Comment by regadas** [comment]
Closed via #2733
